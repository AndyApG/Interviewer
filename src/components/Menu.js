import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'network-canvas-ui';

import { MenuContent } from '.';
import { MenuItem } from './Elements';

/**
  * Renders a menu, updating styles on DOM elements outside of this.
  * @extends Component
  */
class MenuFactory extends Component {
  /**
    * adds listener for key events to close Menu
    */
  componentDidMount() {
    window.onkeydown = this.listenForClose;
  }

  componentDidUpdate() {
    if (this.props.isOpen) {
      document.addEventListener('click', this.outsideClick);
    } else {
      document.removeEventListener('click', this.outsideClick);
    }
  }

  /**
    * removes listener for key events and click events to close Menu
    */
  componentWillUnmount() {
    window.onkeydown = null;
  }

  listenForClose = (e) => {
    const event = e || window.event;

    if (this.props.isOpen && (event.key === 'Escape' || event.keyCode === 27)) {
      this.props.toggleMenu();
    }
  }

  outsideClick = () => {
    this.props.toggleMenu();
  }

  menuClick = (e) => {
    if (e) {
      e.stopPropagation();
      e.preventDefault();
    }
    this.props.toggleMenu();
  }

  // intercepts click events; calls callback and toggles Menu open state
  menuItemClick = (itemClick) => {
    this.menuClick();
    itemClick();
  }

  render() {
    const { children, hideButton, icon, isOpen, items, searchField, title } = this.props;

    const menuItems = items.map(item =>
      (<MenuItem
        key={item.id}
        to={item.to}
        onClick={() => this.menuItemClick(item.onClick)}
        title={item.title}
        isActive={item.isActive}
        interfaceType={item.interfaceType}
        menuType={item.menuType}
      />),
    );

    return (
      <div className="menu" ref={(node) => { this.domNode = node; }}>
        <div className={isOpen ? 'menu__wrap menu__wrap--open' : 'menu__wrap'}>
          <div className="menu__content">
            <MenuContent
              items={menuItems}
              searchField={searchField}
              title={title}
              toggleMenu={this.menuClick}
            />
          </div>
        </div>
        {!hideButton && <div className="menu__burger" onClick={this.menuClick} tabIndex={0} role="menu">
          <Icon name={icon} />
        </div>}
        { children }
      </div>
    );
  } // end render
} // end class

MenuFactory.propTypes = {
  children: PropTypes.any,
  hideButton: PropTypes.bool,
  icon: PropTypes.string,
  isOpen: PropTypes.bool.isRequired,
  items: PropTypes.array,
  title: PropTypes.string,
  toggleMenu: PropTypes.func.isRequired,
  searchField: PropTypes.object,
};

MenuFactory.defaultProps = {
  children: null,
  hideButton: false,
  icon: 'menu',
  items: [],
  searchField: null,
  title: 'Options',
};

export default MenuFactory;
