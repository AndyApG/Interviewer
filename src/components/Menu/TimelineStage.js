import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import timelineImages from '../../images/timeline';

const getTimelineImage = type =>
  get(timelineImages, type, timelineImages.OrdinalBin);

const TimelineStage = props => (
  <div className="timeline-stage">
    <div className="timeline-stage__notch" />
    <div className="timeline-stage__preview">
      <img
        src={getTimelineImage(props.item.type)}
        alt="NameGenerator Interface"
        title="NameGenerator Interface"
      />
    </div>
    <div className="timeline-stage__label">{props.item.label}</div>
  </div>
);

TimelineStage.propTypes = {
  item: PropTypes.object.isRequired,
};

export default TimelineStage;
