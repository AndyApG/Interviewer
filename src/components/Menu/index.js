export { default as Timeline } from './Timeline';
export { default as TimelineStage } from './TimelineStage';
export { default as StatsPanel } from './StatsPanel';
