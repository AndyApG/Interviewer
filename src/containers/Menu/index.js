export { default as MenuContainer } from './MenuContainer';
export { default as SettingsMenu } from './SettingsMenu';
export { default as StagesMenu } from './StagesMenu';
