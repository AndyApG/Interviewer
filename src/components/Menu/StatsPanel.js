import React, { Component } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';
import { Button } from '../../ui/components';
import { actionCreators as uiActions } from '../../ducks/modules/ui';
import { actionCreators as sessionActions } from '../../ducks/modules/session';

class StatsPanel extends Component {
  handleFinishInterview = () => {
    this.props.endSession();
    this.props.closeMenu();
  }

  render() {
    return (
      <div className="stages-menu__stats">
        <h4>Session Information</h4>
        <Button onClick={this.handleFinishInterview}>Finish Interview</Button>
      </div>
    );
  }
};

const mapDispatchToProps = dispatch => ({
  closeMenu: () => dispatch(uiActions.update({ isMenuOpen: false })),
  endSession: () => {
    dispatch(sessionActions.endSession());
    dispatch(push('/'));
  },
});

export { StatsPanel };

export default connect(null, mapDispatchToProps)(StatsPanel);
