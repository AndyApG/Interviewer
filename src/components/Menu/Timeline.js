import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Scroller } from '../../components';
import { TimelineStage } from '.';

/**
  * Renders a menu, updating styles on DOM elements outside of this.
  * @extends Component
  */

// eslint-disable-next-line
 class Timeline extends Component {

  render() {
    const { items } = this.props;
    const menuItems = items.map(item =>
      (<TimelineStage item={item} key={item.id} />),
    );

    return (
      <div className="stages-timeline__timeline">
        <Scroller>
          { menuItems }
        </Scroller>
      </div>
    );
  }
}

Timeline.propTypes = {
  items: PropTypes.array.isRequired,
};

export default Timeline;
