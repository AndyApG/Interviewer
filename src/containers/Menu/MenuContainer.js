import React, { Component } from 'react';
import cx from 'classnames';
import { Button, Icon } from '../../ui/components';
import { StagesMenu, SettingsMenu } from '.';


class MenuContainer extends Component {
  constructor(props) {
    super(props);
    this.state = {
      activePanel: 'stages',
    };

    this.toggleActivePanel = this.toggleActivePanel.bind(this);
  }

  toggleActivePanel() {
    const activePanel = this.state.activePanel === 'settings' ? 'stages' : 'settings';
    this.setState({ activePanel });
  }

  render() {
    return (
      <div className="menu-container">
        <div className="menu-container__content">
          <div className="menu-container__header">
            <Icon name="close" />
          </div>
          <div className="menu-container__panels">
            <div className="menu-panels">
              <div
                className={cx({
                  'menu-panel': true,
                  'menu-panel__settings': true,
                  active: this.state.activePanel === 'settings',
                })}
                onClick={this.toggleActivePanel}
              >
                <SettingsMenu />
              </div>
              <div
                className={cx({
                  'menu-panel': true,
                  'menu-panel__stages': true,
                  active: this.state.activePanel === 'stages',
                })}
                onClick={this.toggleActivePanel}
              >
                <StagesMenu />
              </div>
            </div>
          </div>
          <div className="menu-container__footer">
            <Button color="neon-coral">Return to start screen</Button>
          </div>
        </div>
      </div>
    );
  }
}

export default MenuContainer;
