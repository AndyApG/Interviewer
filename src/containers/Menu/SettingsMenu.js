import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Icon } from '../../ui/components';

// eslint-disable-next-line
class SettingsMenu extends Component {
  render() {
    const { active, onClickInactive } = this.props;
    const handleClickInactive = active ? onClickInactive : null;

    return (
      <React.Fragment>
        <Icon name="settings" />
        <div className="settings-menu" onClick={handleClickInactive}>
          <div className="settings-menu__header">
            <h1>Settings</h1>
          </div>
          <div className="settings-menu__form">
            <p>Some settings and things will go here.</p>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

SettingsMenu.propTypes = {
  active: PropTypes.bool,
  onClickInactive: PropTypes.func,
};

SettingsMenu.defaultProps = {
  active: false,
  onClickInactive: () => {},
};

export default SettingsMenu;
