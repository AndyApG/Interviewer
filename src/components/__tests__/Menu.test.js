/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import Menu from '../Menu';

jest.mock('../../utils/CSSVariables');

const items = [
  { id: 'a', label: 'a title', to: 'a-to' },
  { id: 'b', label: 'b title', to: 'b-to' },
  { id: 'c', label: 'c title', to: 'c-to' },
];

describe('Menu component', () => {
  it('renders menu with list', () => {
    const component = shallow(
      <Menu items={items} isOpen={false} toggleMenu={() => {}} />,
    );

    expect(component).toMatchSnapshot();
  });

  // TODO flush out tests, but the Menu is calling document.getElementById instead of using refs
});
