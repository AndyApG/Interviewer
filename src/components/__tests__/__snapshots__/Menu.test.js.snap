// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component renders menu with list 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MenuFactory
    isOpen={false}
    items={
      Array [
        Object {
          "id": "a",
          "label": "a title",
          "to": "a-to",
        },
        Object {
          "id": "b",
          "label": "b title",
          "to": "b-to",
        },
        Object {
          "id": "c",
          "label": "c title",
          "to": "c-to",
        },
      ]
    }
    searchField={null}
    title="Options"
    toggleMenu={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        null,
        <div
          className="menu__wrap menu__content"
        >
          <Scrollable
            className="menu__scroller"
          >
            <Icon
              className="menu__cross"
              color=""
              name="close"
              onClick={[Function]}
              size="40px"
              style={Object {}}
            />
            <header>
              <h1
                className="menu__title"
              >
                Options
              </h1>
            </header>
            <nav>
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="a title"
                menuType="primary"
                onClick={[Function]}
                to="a-to"
              />
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="b title"
                menuType="primary"
                onClick={[Function]}
                to="b-to"
              />
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="c title"
                menuType="primary"
                onClick={[Function]}
                to="c-to"
              />
            </nav>
          </Scrollable>
        </div>,
      ],
      "className": "menu",
    },
    "ref": [Function],
    "rendered": Array [
      null,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Scrollable
            className="menu__scroller"
          >
            <Icon
              className="menu__cross"
              color=""
              name="close"
              onClick={[Function]}
              size="40px"
              style={Object {}}
            />
            <header>
              <h1
                className="menu__title"
              >
                Options
              </h1>
            </header>
            <nav>
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="a title"
                menuType="primary"
                onClick={[Function]}
                to="a-to"
              />
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="b title"
                menuType="primary"
                onClick={[Function]}
                to="b-to"
              />
              <MenuItem
                icon=""
                interfaceType=""
                isActive={false}
                label="c title"
                menuType="primary"
                onClick={[Function]}
                to="c-to"
              />
            </nav>
          </Scrollable>,
          "className": "menu__wrap menu__content",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <Icon
                className="menu__cross"
                color=""
                name="close"
                onClick={[Function]}
                size="40px"
                style={Object {}}
              />,
              <header>
                <h1
                  className="menu__title"
                >
                  Options
                </h1>
              </header>,
              null,
              <nav>
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="a title"
                  menuType="primary"
                  onClick={[Function]}
                  to="a-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="b title"
                  menuType="primary"
                  onClick={[Function]}
                  to="b-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="c title"
                  menuType="primary"
                  onClick={[Function]}
                  to="c-to"
                />
              </nav>,
            ],
            "className": "menu__scroller",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "menu__cross",
                "color": "",
                "name": "close",
                "onClick": [Function],
                "size": "40px",
                "style": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <h1
                  className="menu__title"
                >
                  Options
                </h1>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Options",
                  "className": "menu__title",
                },
                "ref": null,
                "rendered": "Options",
                "type": "h1",
              },
              "type": "header",
            },
            null,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="a title"
                    menuType="primary"
                    onClick={[Function]}
                    to="a-to"
                  />,
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="b title"
                    menuType="primary"
                    onClick={[Function]}
                    to="b-to"
                  />,
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="c title"
                    menuType="primary"
                    onClick={[Function]}
                    to="c-to"
                  />,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "a",
                  "nodeType": "function",
                  "props": Object {
                    "icon": "",
                    "interfaceType": "",
                    "isActive": false,
                    "label": "a title",
                    "menuType": "primary",
                    "onClick": [Function],
                    "to": "a-to",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "b",
                  "nodeType": "function",
                  "props": Object {
                    "icon": "",
                    "interfaceType": "",
                    "isActive": false,
                    "label": "b title",
                    "menuType": "primary",
                    "onClick": [Function],
                    "to": "b-to",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "c",
                  "nodeType": "function",
                  "props": Object {
                    "icon": "",
                    "interfaceType": "",
                    "isActive": false,
                    "label": "c title",
                    "menuType": "primary",
                    "onClick": [Function],
                    "to": "c-to",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "nav",
            },
          ],
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          null,
          <div
            className="menu__wrap menu__content"
          >
            <Scrollable
              className="menu__scroller"
            >
              <Icon
                className="menu__cross"
                color=""
                name="close"
                onClick={[Function]}
                size="40px"
                style={Object {}}
              />
              <header>
                <h1
                  className="menu__title"
                >
                  Options
                </h1>
              </header>
              <nav>
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="a title"
                  menuType="primary"
                  onClick={[Function]}
                  to="a-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="b title"
                  menuType="primary"
                  onClick={[Function]}
                  to="b-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="c title"
                  menuType="primary"
                  onClick={[Function]}
                  to="c-to"
                />
              </nav>
            </Scrollable>
          </div>,
        ],
        "className": "menu",
      },
      "ref": [Function],
      "rendered": Array [
        null,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Scrollable
              className="menu__scroller"
            >
              <Icon
                className="menu__cross"
                color=""
                name="close"
                onClick={[Function]}
                size="40px"
                style={Object {}}
              />
              <header>
                <h1
                  className="menu__title"
                >
                  Options
                </h1>
              </header>
              <nav>
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="a title"
                  menuType="primary"
                  onClick={[Function]}
                  to="a-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="b title"
                  menuType="primary"
                  onClick={[Function]}
                  to="b-to"
                />
                <MenuItem
                  icon=""
                  interfaceType=""
                  isActive={false}
                  label="c title"
                  menuType="primary"
                  onClick={[Function]}
                  to="c-to"
                />
              </nav>
            </Scrollable>,
            "className": "menu__wrap menu__content",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <Icon
                  className="menu__cross"
                  color=""
                  name="close"
                  onClick={[Function]}
                  size="40px"
                  style={Object {}}
                />,
                <header>
                  <h1
                    className="menu__title"
                  >
                    Options
                  </h1>
                </header>,
                null,
                <nav>
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="a title"
                    menuType="primary"
                    onClick={[Function]}
                    to="a-to"
                  />
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="b title"
                    menuType="primary"
                    onClick={[Function]}
                    to="b-to"
                  />
                  <MenuItem
                    icon=""
                    interfaceType=""
                    isActive={false}
                    label="c title"
                    menuType="primary"
                    onClick={[Function]}
                    to="c-to"
                  />
                </nav>,
              ],
              "className": "menu__scroller",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "className": "menu__cross",
                  "color": "",
                  "name": "close",
                  "onClick": [Function],
                  "size": "40px",
                  "style": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": <h1
                    className="menu__title"
                  >
                    Options
                  </h1>,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Options",
                    "className": "menu__title",
                  },
                  "ref": null,
                  "rendered": "Options",
                  "type": "h1",
                },
                "type": "header",
              },
              null,
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <MenuItem
                      icon=""
                      interfaceType=""
                      isActive={false}
                      label="a title"
                      menuType="primary"
                      onClick={[Function]}
                      to="a-to"
                    />,
                    <MenuItem
                      icon=""
                      interfaceType=""
                      isActive={false}
                      label="b title"
                      menuType="primary"
                      onClick={[Function]}
                      to="b-to"
                    />,
                    <MenuItem
                      icon=""
                      interfaceType=""
                      isActive={false}
                      label="c title"
                      menuType="primary"
                      onClick={[Function]}
                      to="c-to"
                    />,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": "a",
                    "nodeType": "function",
                    "props": Object {
                      "icon": "",
                      "interfaceType": "",
                      "isActive": false,
                      "label": "a title",
                      "menuType": "primary",
                      "onClick": [Function],
                      "to": "a-to",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": "b",
                    "nodeType": "function",
                    "props": Object {
                      "icon": "",
                      "interfaceType": "",
                      "isActive": false,
                      "label": "b title",
                      "menuType": "primary",
                      "onClick": [Function],
                      "to": "b-to",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                  Object {
                    "instance": null,
                    "key": "c",
                    "nodeType": "function",
                    "props": Object {
                      "icon": "",
                      "interfaceType": "",
                      "isActive": false,
                      "label": "c title",
                      "menuType": "primary",
                      "onClick": [Function],
                      "to": "c-to",
                    },
                    "ref": null,
                    "rendered": null,
                    "type": [Function],
                  },
                ],
                "type": "nav",
              },
            ],
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
